proposals/pattern-based-fixed.md

```diff
@@ -0,0 +1,133 @@
+# Pattern-based `fixed` statement
+
+* [x] Proposed
+* [ ] Prototype: Not Started
+* [ ] Implementation: Not Started
+* [ ] Specification: Not Started
+
+## Summary
+[summary]: #summary
+
+Introduce a pattern that would allow types to participate in `fixed` statements. 
+
+## Motivation
+[motivation]: #motivation
+
+The language provides a mechanism for pinning managed data and obtain a native pointer to the underlying buffer.
+
+```C#
+fixed(byte* ptr = byteArray)
+{
+   // ptr is a native pointer to the first element of the array
+   // byteArray is protected from being moved/collected by the GC for the duration of this block 
+}
+
+```
+
+The set of types that can participate in `foreach` is hardcoded and limited to arrays and `System.String`. Hardcoding "special" types does not scale when new primitives such as `ImmutableArray<T>`, `Span<T>`, `Utf8String` are introduced. 
+
+In addition, the current solution for `System.String` relies on a fairly rigid API. The shape of the API implies that `System.String` is a contiguous object that embeds UTF16 encoded data at a fixed offset from the object header. Such approach has been found problematic in several proposals that could require changes to the underlying layout. 
+It would be desirable to be able to switch to something more flexible that decouples `System.String` object from its internal representation for the purpose of unmanaged interop. 
+
+## Detailed design
+[design]: #detailed-design
+
+## *Pattern* ##
+A viable pattern-based �fixed� need to:
+-	Provide the managed references to pin the instance and to initialize the pointer (preferably this is the same reference)
+-	Convey unambiguously the type of the unmanaged element   (i.e. �char� for �string�)
+-	Prescribe the behavior in "empty" case when there is nothing to refer to. 
+-	Should not push API authors toward design decisions that hurt the use of the type outside of `fixed`.
+
+
+I think the above could be satisfied by recognizing a specially named ref-returning member:
+ `ref [readonly] T DangerousGetPinnableReference()`.
+
+In order to be used by the `foreach` statement the following conditions must be met:
+
+1)	There is only one such member provided for a type.
+1)	Returns by `ref` or `ref readonly`. 
+(readonly is permitted so that authors of immutable/readonly types could implement the pattern)
+1)  T is an unmanaged type.
+(since `T*` becomes the pointer type)
+1)	Returns managed `nullptr` when there is no data to pin � probably the cheapest way to convey emptiness.
+(note that �� string returns a ref to �\0� since strings are zero-terminated)
+
+
+Alternatively for the `#3` we can allow the result in empty cases be undefined or implementation-specific. 
+That, however, may make the API more dangerous and prone to abuse and unintended compatibility burdens. 
```

Comment:
I think it needs to be defined.  We left it as undefined for a while for arrays, and that just led to writing less efficient code, since the call sites would then end up duplicating checks that were emitted by the compiler anyway.

Reply:
Yes, "undefined" often leads to "defined in tribal knowledge" situation. 
Implementer still ends up bound by compat, while users need to program defensively.

I have added this as a point to discuss since we did "undefined" in this area before.  
The spec for current cases of "fixed" still says "undefined" although the behavior is well known and hardly can be changed.

Reply:
> The spec for current cases of "fixed" still says "undefined"

We'd agreed to update the spec here:
https://github.com/dotnet/roslyn/issues/6707
and did so here:
https://github.com/dotnet/csharplang/pull/16
though I don't know if those changes have propagated anywhere "real".